### PR TITLE
[Feature] workflow level task dry run

### PR DESCRIPTION
**Pull request by andream7**
## Purpose of the pull request

This pull request adds workflow level task dry run.

## Brief change log

https://github.com/andream7/summer2021-week-report/blob/main/mid-report.md

## Verify this pull reques

This change added tests and can be verified as follows:

  - modified ExecutorService2Test and ProcessServiceTest.

### CODE REVIEW

File: dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/Constants.java [hunk 1]
```diff
@@ -1083,4 +1083,11 @@ private Constants() {
     public static final String TASK_DEPENDENCE_PROJECT_NAME = "projectName";
     public static final String TASK_DEPENDENCE_DEFINITION_ID = "definitionId";
     public static final String TASK_DEPENDENCE_DEFINITION_NAME = "definitionName";
+
+    /**
+     * dry run state
+     */
+    public static final Integer NORMAL_STATE = 0;
+    public static final Integer DRY_RUN_STATE = 1;
```
Comment (wen-hemin) [inline dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/Constants.java]: DRY_RUN_FLAG_NO and DRY_RUN_FLAG_YES
'state' change to 'flag'
Integer change to int

File: dolphinscheduler-dao/src/main/java/org/apache/dolphinscheduler/dao/entity/Command.java [hunk 2]
```diff
@@ -357,6 +373,7 @@ public int hashCode() {
         result = 31 * result + (processInstancePriority != null ? processInstancePriority.hashCode() : 0);
         result = 31 * result + (updateTime != null ? updateTime.hashCode() : 0);
         result = 31 * result + (workerGroup != null ? workerGroup.hashCode() : 0);
+        result = 31 * result + (dryRun != null ? dryRun.hashCode() : 0);
         result = 31 * result + (environmentCode != null ? environmentCode.hashCode() : 0);
```
Comment (wen-hemin) [inline dolphinscheduler-dao/src/main/java/org/apache/dolphinscheduler/dao/entity/Command.java]: Adjust the order

File: dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/runner/TaskExecuteThread.java [hunk 3]
```diff
@@ -191,14 +197,25 @@ public void run() {
...
             //init varPool
             this.task.getParameters().setVarPool(taskExecutionContext.getVarPool());
-            // task handle
-            this.task.handle();
 
-            // task result process
-            if (this.task.getNeedAlert()) {
-                sendAlert(this.task.getTaskAlertInfo());
+            taskInstanceMapper = SpringApplicationContext.getBean(TaskInstanceMapper.class);
+            TaskInstance taskInstance = taskInstanceMapper.selectById(taskExecutionContext.getTaskInstanceId());
```
Comment (wen-hemin) [inline dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/runner/TaskExecuteThread.java]: You can use the method "ProcessService.findTaskInstanceById()", no need introduce the TaskInstanceMapper class

File: sql/dolphinscheduler_mysql.sql [hunk 4]
```diff
@@ -332,6 +332,7 @@ CREATE TABLE `t_ds_command` (
   `process_instance_priority` int(11) DEFAULT NULL COMMENT 'process instance priority: 0 Highest,1 High,2 Medium,3 Low,4 Lowest',
   `worker_group` varchar(64)  COMMENT 'worker group',
   `environment_code` bigint(20) DEFAULT '-1' COMMENT 'environment code',
+  `dry_run` int NULL DEFAULT 0 COMMENT 'dry run state：0 normal, 1 dry run',
```
Comment (wen-hemin) [inline sql/dolphinscheduler_mysql.sql]: 'state' change to 'flag'

File: dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/runner/TaskExecuteThread.java [hunk 5]
```diff
@@ -191,14 +197,25 @@ public void run() {
             this.task.init();
             //init varPool
             this.task.getParameters().setVarPool(taskExecutionContext.getVarPool());
```
Comment (wen-hemin) [inline dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/runner/TaskExecuteThread.java:204]: Does the previous code need to be executed when dry run?

**Review by wen-hemin — CHANGES_REQUESTED** [review]
Please check

File: sql/dolphinscheduler_mysql.sql (hunk 4 above)
Comment (andream7) [reply sql/dolphinscheduler_mysql.sql]: ok

File: dolphinscheduler-dao/src/main/java/org/apache/dolphinscheduler/dao/entity/ErrorCommand.java [hunk 6]
```diff
@@ -120,7 +120,12 @@
      */
     private Long environmentCode;
 
-    public ErrorCommand(){}
+    /**
+     * dry run state
```
Comment (wen-hemin) [inline dolphinscheduler-dao/src/main/java/org/apache/dolphinscheduler/dao/entity/ErrorCommand.java]: fix comment

**Review by wen-hemin — COMMENTED** [review]
Please check

**Comment by CalvinKirs** [comment]
PTAL,Thx @wen-hemin

**Review by wen-hemin — APPROVED** [review]
+1

File: sql/dolphinscheduler_postgre.sql [hunk 7]
```diff
@@ -297,6 +298,7 @@ CREATE TABLE t_ds_error_command (
   worker_group varchar(64),
   environment_code bigint DEFAULT '-1',
   message text ,
+  dry_ru int DEFAULT '0' ,
```
Comment (kezhenxu94) [inline sql/dolphinscheduler_postgre.sql:301]: Please also add this field to `sql/dolphinscheduler_h2.sql`

**Pull request merged**